### ISSUE TITLE
Forest Reserves: List of additional features required for the interactive map of each Forest Reserve record (Fields)

### ISSUE THREAD

**Issue by ygoex** [National-Forestry-Authority/brms#19]
For each CFR record, additional data fields as follows are required:

- Status of boundary records in the system (work required, work-in-progress, complete, under review) for whole CFR. It a select list, and we use a taxonomy.
- Priority for action (Urgent, high, medium, low) for whole CFR. It is a select list, and we use a taxonomy.
- Area (ha)
- Total boundary length (km)
- Total cutline length (km)
- Total natural boundary length (km)
- Year Gazetted, just the year.
- Boundary plan Nos, short text field.
- Topo Map Serial Nos, short text field.
- Cadastral Sheet No, short text field.
- Status of boundary description of the ground (*surveyed, surveyed and marked, marked but not recently checked*)
- If marked as surveyed, expose a multi-field survey details form that has the following subfields:
   - Year
   - Surveyor name (mandatory)
   - Surveyor username (optional)
   - Description of survey work (mandatory), format long text.
   - List of ordered way points (with / without pillars) --> see above
- Log of actions taken (dated / time-stamped)

**Comment by ygoex** [comment]
- This requires some UI/UX for displaying the fields.
- We should use Paragraphs for this task.

**Comment by ygoex** [comment]
We need @LeScruf to review the fields in the description.

**Comment by LeScruf** [comment]
@senyonjo-edward @ygoex - Here is a merged list of fields for the CFR records. @senyonjo-edward please review / edit as needed. Thx!

**Comment by senyonjo-edward** [comment]
@ygoex
For each CFR and LFR record, additional data fields as follows are required:
•	Forest Reserve Type (Dropdown CFR, LFR) only two options possible
•	Serial number (Dropdown e.g. BN/3)
•	Name of the Forest Reserve (Dropdown text)
•	District (Dropdown text
•	County (text)
•	Map sheet reference (Dropdown e.g. 56/3). Some forest reserves have more than one map reference. (short text field)
•	Cadastral Sheet No 1:50,000 scale, short text field.( Dropdown)
•	Cadastral Sheet No 1:10,000 scale, short text field.( Dropdown)
•	Boundary Plan No., short text field. Dropdown (some forest reserves have more than one BP number)
•	Land Registration No., short text field. (PL, CL, FRV, SE). CFRs can have more than one of these types or a combination of them)
•	Gazetted Area (ha) (Fixed as per the SI document)
•	Mapped Area (ha)
•	Total boundary length (km)
o	Total cutline length (km)
o	natural boundary length (km)
	Riverline (km)
	Shoreline (km)
	Protected area (km)
•	Year Gazetted, just the year. (dropdown)
•	Other Gazette mentions (Dropdown-1932, 1948, 1952,1969,1998)
•	Status of boundary records in the system (work required, work-in-progress, under review, complete, work stalled) for the whole CFR. It is a select list, and we use a taxonomy.
•	Priority for action (Urgent, high, medium, low, impromptu) for the whole CFR. It is a select list, and we use a taxonomy.
•	Status of boundary description of the ground (surveyed, surveyed and marked, marked but not recently checked)
•	If marked as surveyed, expose a multi-field survey details form that has the following subfields:
o	Year
o	Surveyor name (dropdown) mandatory)
o	Surveyor username (optional)
o	Description of survey work (mandatory), format long text.
o	List of ordered waypoints (with/without pillars) --> see above
•	Log of actions taken (dated/time-stamped)
**NB: Please also refer to the Statutory Instrument attached/shared**

**Comment by senyonjo-edward** [comment]
@ygoex 
I have sent the Statutory Instrument (SI) via email.

**Comment by senyonjo-edward** [comment]
Dear Yolanda,

Please find attached the Statutory Instrument legal document that I
mentioned on issue #
<https://github.com/National-Forestry-Authority/brms/issues/19>19 on the
BRMS forum.

Thank you.
Edward Ssenyonjo
Coordinator Inventory & Surveys, National Forestry Authority
*Address:* Plot 10/20, spring road, P. O. Box, 70863, Kampala
*Email:* ***@***.*** | ***@***.***
*Tel:* +256 786 461 620 *CG* | +256 772 521 432
*skype:* senyonjo_edward

**Comment by LeScruf** [comment]
Thank you Edward!

**Comment by ygoex** [comment]
Dear Edward,

I’m afraid I’ve not been able to download the Statutory Instrument legal document, it doesn’t seem to be attached, or it’s gone missing for no apparent reason.

Could you please send it again so I can pass it to the rest of the team.

Many thanks,
Yolanda

**Comment by peacog** [comment]
Attaching the Statutory Instrument document.

[Statutory Instrument 1998.pdf](https://github.com/National-Forestry-Authority/brms/files/10921183/Statutory.Instrument.1998.pdf)

**Comment by peacog** [comment]
Here is a preliminary attempt to define the Drupal content type for CFRs and LFRs. Let's review it and fill in the gaps in the next BRMS meeting.

### Content type
We currently have a content type CFR Master Maps (machine name central_forest_reserve)

Q. Should we remove CFR Master Maps and replace it with a Forest Reserve (forest_reserve) content that has a Forest Reserve Type field with two possible values?
  - Central Forest Reserve (CFR)
  - Local Forest Reserve (LFR)

Q. Edward specified some fields as dropdown text, the values to be taken from the [SI document](https://github.com/National-Forestry-Authority/brms/files/10921183/Statutory.Instrument.1998.pdf). Should these be List (text) or taxonomy reference fields? Or can they be plain text fields? 

Q. After the initial import of CFRs and LFRS is it likely that new reserves will be added manually in the future?

### Forest Reserve content type fields

| Field                          | Drupal Field Type      | Required | Multi-value | Read-only | Comment                                                                                                                                                                            |
|--------------------------------|------------------------|----------|-------------|-----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Forest reserve type            | List (text)            | Yes      |             |           | Possible values: CFR, LFR                                                                                                                                                          |
| Name                           | Title                  | Yes      |             |           | Do we need to validate this against a list of names derived from [SI document](https://github.com/National-Forestry-Authority/brms/files/10921183/Statutory.Instrument.1998.pdf) ? |
| Serial number                  | ?                      |          |             |           | Should this be a Text or List (text) field? Valid values are in [SI document](https://github.com/National-Forestry-Authority/brms/files/10921183/Statutory.Instrument.1998.pdf).   |
| District                       | List (text)            |          |             |           | Should we use taxonomy for this so that we can filter by District in views?                                                                                                        |
| County                         | Text (plain)           |          |             |           | Should we use taxonomy?                                                                                                                                                            |
| Map sheet reference            | ?                      |          | Yes         |           | Should we use taxonomy?                                                                                                                                                            |
| Cadastral Sheet No 1:50,000    | ?                      |          |             |           |                                                                                                                                                                                    |
| Cadastral Sheet No 1:10,000    | ?                      |          |             |           |                                                                                                                                                                                    |
| Boundary Plan No               | ?                      |          | Yes         |           |                                                                                                                                                                                    |
| Land Registration No           | ?                      |          | Yes         |           |                                                                                                                                                                                    |
| Gazetted Area (ha)             | Number (decimal)       |          |             | Yes       | Imported from SI data.                                                                                                                                                             |
| Mapped Area (ha)               | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Total boundary length (km)     | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Total cutline length (km)      | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Natural boundary length (km)   | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Riverline (km)                 | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Shoreline (km)                 | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Protected area (km)            | Number (decimal)       |          |             |           |                                                                                                                                                                                    |
| Year Gazetted                  | Year only              |          |             |           |                                                                                                                                                                                    |
| Other Gazette mentions         | Year only              |          | Yes         |           |                                                                                                                                                                                    |
| Status of boundary records     | Taxonomy (select)      |          |             |           |                                                                                                                                                                                    |
| Priority for action            | Taxonomy (select)      |          |             |           |                                                                                                                                                                                    |
| Status of boundary description | Taxonomy (select)      |          |             |           |                                                                                                                                                                                    |
| Year surveyed                  | Year only              |          |             |           | Conditional field available when Status of boundary description is 'surveyed'                                                                                                      |
| Surveyor name                  | ?                      | Yes      |             |           | Conditional field available when Status of boundary description is 'surveyed'                                                                                                      |
| Surveyor username              | User                   |          |             |           | Conditional field available when Status of boundary description is 'surveyed'                                                                                                      |
| Description of survey work     | Text (formatted, long) | Yes      |             |           | Conditional field available when Status of boundary description is 'surveyed'                                                                                                      |
| List of ordered waypoints      | ?                      |          |             |           | Conditional field available when Status of boundary description is 'surveyed'                                                                                                      |
| Log of actions                 | ?                      |          |             |           |                                                                                                                                                                                    |

### Taxonomy vocabularies
- Boundary records status
- Priority 
- Boundary description status

**Comment by LeScruf** [comment]
Hi @peacog & @senyonjo-edward, I answered three questions in bold above - the table questions & comments we can quickly go through today?

**Comment by ygoex** [comment]
Once the table is complete, we are going to create a CVS for NFA.

**Comment by ygoex** [comment]
Protected area, shoreline, riverline need to be added to Natural boundary

**Comment by ygoex** [comment]
We create blocked users for surveyors that are not part of the NFA any more, we create a fullname for the user entity and we relate a surveyor by the fullname.

**Comment by ygoex** [comment]
List of ordered waypoints: one field that uploads a kml and displays a table. This should be another GeoLayer entity reference.

**Comment by ygoex** [comment]
Log of actions has been already done here: https://github.com/National-Forestry-Authority/brms/issues/21

**Comment by ygoex** [comment]
We would need to cross check the fields:

**Comment by LeScruf** [comment]
@ygoex I have gone through and crossed checked the above descriptors with those in the table prepared by Edward, and they have all already be incorporated as needed. Thks!

**Comment by ygoex** [comment]
We need to add a field to hold the CFR identifier.

**Comment by LeScruf** [comment]
@senyonjo-edward - Edward can we use the CFR identifiers from the Statutory Instrument for this? We can discuss?

**Comment by senyonjo-edward** [comment]
Hi Andrew,

@ygoex We should use the FD code (Forest Reserve serial number) as the unique forest reserve identifier.

However, we could also add a column with sequential serial numbering of the forest reserves

**Comment by LeScruf** [comment]
Hi @peacog , thanks for this. Edward and I will review as soon as we can. In the meantime, we think that the first tab 'CFRs' on https://docs.google.com/spreadsheets/d/1UxIPh0a-ciR-56PHblqVpwa9IMtUj3mR82WeedmU9dw/edit#gid=162492019 is getting close. You will note that the structure of this has evolved - and so it would good to also evolve the form correspondingly. This does not stop us from testing the form of course - which I hope can be done very shortly. But I think its looking good! :)

**Comment by peacog** [comment]
Moving back to In progress to:

- [ ] synchronise the form with the changes made to the spreadsheet.
- [ ] improve usability of form by organising field groups into vertical tabs

**Comment by LeScruf** [comment]
@peacog @ygoex @senyonjo-edward - the open layers map is potentially going to start getting cluttered - we need to figure out a way of how to easily keep track of different polygon layers (which is being done through layer control) - figuring out a line and colour scheme that makes sense and provides a good UI.

**Comment by peacog** [comment]
Hi @senyonjo-edward @LeScruf . I've added a new taxonomy field to the Geloayer that lets you choose a layer style that controls the colour and style of the polygon for that layer.

You can add styles here: https://brms.stg.envs.utils.nfa.go.ug/admin/structure/taxonomy/manage/layer_style/overview
And you can assign the style to a layer in the Forest Reserve node, for example https://brms.stg.envs.utils.nfa.go.ug/node/26139/edit

I've set a predefined list of colours, and lines can be solid, dotted, or dashed. If you want more flexibility we could change the taxonomy settings to allow you to choose any colour you want, and/or different dashed line styles. Test it when you have a change and let me know if you have any suggestions.

**Comment by peacog** [comment]
Hi @LeScruf - did you have a chance to look at the geolayer styles described in the previous comment? Let me know thoughts

**Comment by LeScruf** [comment]
hi @peacog - this is looking very good, thank you - a question is how is this exposed in a way which is straight-forward for a NFA admin to manage? Not immediately pressing - but perhaps one of several functionality areas that can be grouped in an NFA admin page view to be easily administered and adjusted going forward? Curious - did you code this up, or is there a per-existing module that does this?

**Comment by peacog** [comment]
It's a custom implementation @LeScruf . I created a new taxonomy vocabulary called Layer style and added color, line width and line style fields to it. A user with permission to add taxonomy terms can create layer styles here: https://brms.stg.envs.utils.nfa.go.ug/nfa-landing?destination=/admin/structure/taxonomy/manage/layer_style/overview i.e. by going to Structure > Taxonomy > Layer style
![image](https://github.com/National-Forestry-Authority/brms/assets/1312689/913e3fdb-9bc1-4dc2-a00d-6e7467306e14)
